source/common/upstream/cluster_manager_impl.cc

```diff
@@ -473,18 +473,45 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
     cluster_warming_cb(cluster_name, ClusterWarmingState::Starting);
     cluster_entry->cluster_->initialize([this, cluster_name, cluster_warming_cb] {
       auto warming_it = warming_clusters_.find(cluster_name);
-      auto& cluster_entry = *warming_it->second;
+      auto& warming_cluster_entry = *warming_it->second;
 
       // If the cluster is being updated, we need to cancel any pending merged updates.
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If onConfigUpdate was triggered by an EDS update that had no references to this
```

Comment:
I think I'm still missing something here. Why isn't the management server responding with the hosts that the Envoy should have? Can you potentially add more comments or we can discuss in this comment thread until we are all on the same page?

Reply:
In reading through https://github.com/envoyproxy/envoy/issues/5168 I'm wondering why this isn't a management server issue. Can you just not have your management server enter the LB pool until it's fully ready to serve requests?

Reply:
I think the best way to think this through is with a concrete example. Let's imaging we're switching from port 80 to 443 (EDS) and HTTP to HTTPS to backend (CDS).

@mattklein123 are you concerned that we _never_ see the 443 EDS host update? Or that we temporarily won't see it during config update?

Reply:
An example would be great. My concern here is that it's not clear to me that this is an Envoy issue vs. a management server issue. I.e., why is the management sever accepting connections if it can't actually serve them? Is there a chance that this workaround is going to change legitimate behavior such as the management server wanting to remove all hosts when a cluster has been refreshed, etc. 

I think I would mainly just like to better understand why we think this is the right fix to do here.

Reply:
My thinking is that the management server always has the capability to send an empty host update (or the 443 hosts in the above example). There may be a period of inconsistency, but arguably that's inevitable when not doing make-before-break consistency.

Reply:
I think I'm probably missing something fundamental, so apologies, but why would the the management server send an empty update if there are hosts?

Reply:
This is the example you gave in:

> Is there a chance that this workaround is going to change legitimate behavior such as the management server wanting to remove all hosts when a cluster has been refreshed, etc.

Reply:
@mattklein123  just to be clear, management server is not sending empty update for this cluster if there are hosts. It sent ClusterLoadAssignment response for some other cluster (there is no reference to this cluster in the response) and we call `onConfigUpdate` for this cluster from here with empty because we did not find any resources for this cluster https://github.com/envoyproxy/envoy/blob/master/source/common/config/grpc_mux_impl.cc#L171
Normally this is fine because if it is initialized we just increment the stat and onPreInit would do nothing. But if it is in the initializing(warming) state, it is clearing the clusters.

Reply:
Added more docs. See if that makes sense.